From 593cba04174e474e45fd95af0116d9d5c4bcda67 Mon Sep 17 00:00:00 2001
From: Hongbo Yao <yaohongbo@huawei.com>
Date: Fri, 26 Feb 2021 20:21:25 +0800
Subject: [PATCH] mpam: fix potential resource leak in mpam_domains_init
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 593cba04174e474e45fd95af0116d9d5c4bcda67


hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

Variable "d" going out of scope leaks the storage it points to.

Signed-off-by: Hongbo Yao <yaohongbo@huawei.com>
Reviewed-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/mpam.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kernel/mpam.c b/arch/arm64/kernel/mpam.c
index d18868d0b296..eb9e38284bc7 100644
--- a/arch/arm64/kernel/mpam.c
+++ b/arch/arm64/kernel/mpam.c
@@ -1219,8 +1219,10 @@ static void mpam_domains_init(struct resctrl_resource *r)
 		d->cpus_list = n->cpus_list;
 
 		d->ctrl_val = kmalloc_array(rr->num_partid, sizeof(*d->ctrl_val), GFP_KERNEL);
-		if (!d->ctrl_val)
+		if (!d->ctrl_val) {
+			kfree(d);
 			return;
+		}
 
 		list_add_tail(&d->list, add_pos);
 
-- 
2.26.2

